### PR TITLE
Feature/multipart uploads

### PR DESCRIPTION
**Pull request by shabbyrobe**
~This is not ready at all. I've not even seen it run yet much less work!~ (No longer applicable!) I've opened the PR to open the discussion though, because it's another fairly massive PR with a lot of new functionality, a lot of which is loaded with compromise.

I've focused on getting a first pass of this in here that doesn't make use of any backends and keeps everything in memory. I've put all the logic for multipart handling in a separate struct because it's quite a messy API generally and as we talked about earlier, the `GoFakeS3` struct is getting a little large. My separation here is somewhat arbitrary though: `GoFakeS3` is still used for all the HTTP-related stuff.

The comment on `uploader` describes the current limitations:
```
// uploader manages multipart uploads.
//
// Multipart upload support has the following rather severe limitations (which
// will hopefully be addressed in the future):
//
//	- uploads do not interface with the Backend, so they do not
// 	  currently persist across reboots
//
//	- upload parts are held in memory, so if you want to upload something huge
//	  in multiple parts (which is pretty much exactly what you'd want multipart
//	  uploads for), you'll need to make sure your memory is also sufficiently
//	  huge!
//
// At this stage, the current thinking would be to add a second optional
// Backend interface that allows persistent operations on multipart upload
// data, and if a Backend does not implement it, this limited in-memory
// behaviour can be the fallback. If that can be made to work, it would provide
// good convenience for Backend implementers if their use case did not require
// persistent multipart upload handling.
```

If you're comfortable with me pressing on, I'd like to leave addressing those limitations for a later date to try to reduce the scope of this round of work, which is already quite large.

I've brought in an additional dependency as well, which I'm always reluctant to do, but Go does not provide a sorted map implentation. I've used https://github.com/ryszard/goskiplist, which I've used before with good results.

I'll start focusing on tests now, and finding all the parts where I read the documentation wrong!

### CODE REVIEW

**Comment by shabbyrobe** [comment]
```
- Coverage    64.4%   39.51%   -24.89%  
```
Ah oops! It won't stay that way though :wink:

**Comment by shabbyrobe** [comment]
Phew! This one has been rather hard, but I've finally got this to a point where it works with `awscli`.

I've improved the test coverage of multipart uploads, but I think I'd like to draw a line under it here if you're comfortable. I'd like to clean the tester up a bit as I have some ideas for how to keep the `testServer` struct from ballooning any worse than it is, and like we talked about earlier I'd also like to get an assertion library in there too. I'd like to tackle that as a separate piece of work though as it could be pretty fussy stuff; tangling it up with this PR would bloat it well beyond its current level of bloat!

I'd like to add some more combinatorial coverage for things like the markers, prefixes and limits soon too but I think we need some sort of small tool for writing tests against S3 itself that prove the assumptions gofakes3 makes about its behaviour in corner cases are correct. In this PR, you can see I've hit a bit of a limit there with that commented-out test case in `TestListMultipartUploadsPrefix`.

If you're comfortable, would you mind if we declared this "done enough to be better than nothing" soon, with a view to improving the depth and cleanliness of the test coverage as that becomes easier?

**Comment by johannesboyne** [comment]
I'll try to carve out a bit of time later to review it ;) thanks for the hard work!

File: validation.go [hunk 1]
```diff
@@ -47,3 +47,18 @@ func ValidateBucketName(name string) error {
 
 	return nil
 }
+
+func validNumericString(v string) bool {
```
Comment (johannesboyne) [inline validation.go]: Not sure where this is currently used, but I assume it will be as well relevant later? ;-)

Comment (shabbyrobe) [reply validation.go]: Oh, actually I think this is a remnant from a previous experiment. I think it can probably be culled, but maybe it might be useful? Not sure. Would you like me to remove it?

File: gofakes3.go [hunk 2]
```diff
@@ -22,20 +26,41 @@ const (
...
+	// 2KiB works, but Amazon uses 2KB...  that's a much worse time to discover
+	// the disparity!
 	DefaultMetadataSizeLimit = 2000
 
+	// Like DefaultMetadataSizeLimit, the docs don't specify MB or MiB, so we
+	// will accept 5MB for now. The Go client SDK rejects 5MB with the error
+	// "part size must be at least 5242880 bytes", which is a hint that it
+	// has been interpreted as MiB at least _somewhere_, but we should remain
+	// liberal in what we accept in the face of ambiguity.
+	DefaultUploadPartSize = 5 * 1000 * 1000
```
[the remark this answers is not in the document]
Comment (shabbyrobe) [reply gofakes3.go]: For sure, that's a great idea. I'll do it next time I add or change an option.

File: error.go [hunk 3]
```diff
@@ -45,16 +45,35 @@ const (
...
+	ErrInvalidPartOrder ErrorCode = "InvalidPartOrder"
+
+	ErrInvalidURI ErrorCode = "InvalidURI"
+
+	ErrMetadataTooLarge ErrorCode = "MetadataTooLarge"
+	ErrMethodNotAllowed ErrorCode = "MethodNotAllowed"
+	ErrMalformedXML     ErrorCode = "MalformedXML"
+
+	// You must provide the Content-Length HTTP header.
+	ErrMissingContentLength ErrorCode = "MissingContentLength"
```
[the remark this answers is not in the document]
Comment (shabbyrobe) [reply error.go]: I must confess I've been cheating 😆. I've basically been copying/pasting the stuff from here wholesale:
https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html

**Comment by shabbyrobe** [comment]
Awesome, thanks heaps for looking into this one for me. It's a bit of a beast, I know. I think I will need to extend it to hit the backend in some capacity eventually as this implementation limits your maximum upload to the virtual memory you have available. Unfortunately, this is actually not enough for my employer's project; some of the files are 20gb+! I just wanted to get a first pass of it in so the diff didn't get any crazier.
